preview/Win7Msix/Win7MSIXInstaller/StartupTask.cpp

```diff
@@ -0,0 +1,193 @@
+        RETURN_IF_FAILED(msixCoreFolder->RegisterTaskDefinition(taskNameBstr, taskDefinition.Get(), TASK_CREATE_OR_UPDATE,
+            variantNull /*userId*/, variantNull /*password*/, TASK_LOGON_INTERACTIVE_TOKEN, variantNull /*sddl*/, &registeredTask));
+    }
+    return S_OK;
+}
+
+HRESULT StartupTask::ExecuteForRemoveRequest()
+{
+    if (m_executable.empty())
+    {
+        return S_OK;
+    }
+
+    AutoCoInitialize coInit;
+    RETURN_IF_FAILED(coInit.Initialize(COINIT_MULTITHREADED));
+    {
+        // New scope so that ComPtrs are released prior to calling CoUninitialize
+
+        VARIANT variantNull;
+        ZeroMemory(&variantNull, sizeof(VARIANT));
+        variantNull.vt = VT_NULL;
+
+        ComPtr<ITaskService> taskService;
+        RETURN_IF_FAILED(CoCreateInstance(CLSID_TaskScheduler,
+            NULL,
+            CLSCTX_INPROC_SERVER,
+            IID_ITaskService,
+            reinterpret_cast<void**>(&taskService)));
+        RETURN_IF_FAILED(taskService->Connect(variantNull, variantNull, variantNull, variantNull));
+
+        Bstr rootFolderPathBstr(windowsTaskFolderName);
+        ComPtr<ITaskFolder> rootFolder;
+        RETURN_IF_FAILED(taskService->GetFolder(rootFolderPathBstr, &rootFolder));
+
+        Bstr taskFolderBstr(taskFolderName);
+        ComPtr<ITaskFolder> msixCoreFolder;
+        RETURN_IF_FAILED(rootFolder->GetFolder(taskFolderBstr, &msixCoreFolder));
+
+        Bstr taskNameBstr(m_msixRequest->GetPackageInfo()->GetPackageFullName());
+        ComPtr<IRegisteredTask> registeredTask;
+        RETURN_IF_FAILED(msixCoreFolder->DeleteTask(taskNameBstr, 0 /*flags*/));
+    }
+    return S_OK;
+}
+
+HRESULT StartupTask::ParseManifest()
+{
+    ComPtr<IMsixDocumentElement> domElement;
+    RETURN_IF_FAILED(m_msixRequest->GetPackageInfo()->GetManifestReader()->QueryInterface(UuidOfImpl<IMsixDocumentElement>::iid, reinterpret_cast<void**>(&domElement)));
+
+    ComPtr<IMsixElement> element;
+    RETURN_IF_FAILED(domElement->GetDocumentElement(&element));
+
+    ComPtr<IMsixElementEnumerator> extensionEnum;
+    RETURN_IF_FAILED(element->GetElements(extensionQuery.c_str(), &extensionEnum));
+    BOOL hasCurrent = FALSE;
+    RETURN_IF_FAILED(extensionEnum->GetHasCurrent(&hasCurrent));
+    while (hasCurrent)
+    {
+        ComPtr<IMsixElement> extensionElement;
+        RETURN_IF_FAILED(extensionEnum->GetCurrent(&extensionElement));
+        Text<wchar_t> extensionCategory;
+        RETURN_IF_FAILED(extensionElement->GetAttributeValue(categoryAttribute.c_str(), &extensionCategory));
+
+        if (wcscmp(extensionCategory.Get(), startupTaskCategoryNameInManifest.c_str()) == 0)
+        {
+            Text<wchar_t> executable;
+            RETURN_IF_FAILED(extensionElement->GetAttributeValue(executableAttribute.c_str(), &executable));
+            m_executable = m_msixRequest->GetFilePathMappings()->GetExecutablePath(executable.Get(), m_msixRequest->GetPackageInfo()->GetPackageFullName().c_str());

```

Comment:
Does schema validation guarantee that there is exactly one of these defined?

Reply:
good catch, handling multiple of these

---
In reply to: [277072724](https://github.com/Microsoft/msix-packaging/pull/87#discussion_r277072724) [](ancestors = 277072724)